### PR TITLE
librbd: Add a function to list image watchers

### PR DESCRIPTION
**Pull request by adamwg**
Currently the only way to see an image's watchers is the `rbd status` command. Add a function to the librbd API that provides the same information, and refactor `rbd status` to use the API call.

### CODE REVIEW

File: src/librbd/internal.cc [hunk 1]
```diff
@@ -2356,7 +2356,35 @@ bool compare_by_name(const child_info_t& c1, const child_info_t& c2)
...
 
+  int list_watchers(ImageCtx *ictx,
+		    std::list<librbd::image_watcher_t> &watchers)
+  {
+    int r;
+    std::string header_oid;
+    std::list<obj_watch_t> obj_watchers;
+
+    if (ictx->old_format) {
+      header_oid = ictx->name + RBD_SUFFIX;
```
Comment (dillaman) [inline src/librbd/internal.cc]: Nit: use ```util::old_header_name```

File: src/librbd/internal.cc [hunk 2]
```diff
@@ -2356,7 +2356,35 @@ bool compare_by_name(const child_info_t& c1, const child_info_t& c2)
...
+    } else {
+      header_oid = RBD_HEADER_PREFIX + ictx->id;
```
Comment (dillaman) [inline src/librbd/internal.cc]: Nit: use ```util::header_name```

File: src/librbd/internal.cc (hunk 1 above)
Comment (adamwg) [reply src/librbd/internal.cc]: Thanks! Was unaware of these util functions. Updated.

**Comment by adamwg** [comment]
@dillaman Done. I'm not super confident in my cython skills, but it appears to build. Is it easy for me to run the tests in src/test/pybind myself, or is there a CI check that will take care of that?

File: src/test/pybind/test_rbd.py [hunk 3]
```diff
@@ -854,6 +854,12 @@ def test_metadata(self):
             metadata = list(self.image.metadata_list())
             eq(len(metadata), N - i - 1)
 
+    def test_watchers_list(self):
+        watchers = list(self.image.watchers_list())
+        # The image is open, in r/w mode, from a previous test, so expect there
```
Comment (dillaman) [inline src/test/pybind/test_rbd.py]: Nit: the image is opened as part of the test setup (https://github.com/ceph/ceph/pull/19188/commits/1e5c0daaca778694e6f4dfd73f59e4095cedd27e#diff-1e9029bdf63884f2975c4a2bd8fab90aR300) and not part of a previous test

Comment (adamwg) [reply src/test/pybind/test_rbd.py]: Ah, good catch. I'm used to the go bindings, where creating an `Image` struct is a separate step from opening it, so I assumed the `Image` creation in the setup was not what did the open. I've updated the comment.

Trying to run the tests fails for me. I think it's looking for a running ceph cluster that I don't have configured locally:

```console
% nosetests -v test_rbd -e '.*many_snaps'
unable to get monitor info from DNS SRV with service name: ceph-mon
no monitors specified to connect to.
ERROR

======================================================================
ERROR: test suite for <module 'test_rbd' from '/home/awg/ceph/src/test/pybind/test_rbd.pyc'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/suite.py", line 209, in run
    self.setUp()
  File "/usr/lib/python2.7/dist-packages/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/usr/lib/python2.7/dist-packages/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/usr/lib/python2.7/dist-packages/nose/util.py", line 471, in try_run
    return func()
  File "/home/awg/ceph/src/test/pybind/test_rbd.py", line 39, in setup_module
    rados.connect()
  File "rados.pyx", line 878, in rados.Rados.connect (/home/awg/ceph/build/src/pybind/rados/rados.c:11665)
ObjectNotFound: [errno 2] error connecting to the cluster

----------------------------------------------------------------------
Ran 0 tests in 0.187s

FAILED (errors=1)```

If you can give the python bits a test that would be great.

**Comment by adamwg** [comment]
jenkins: retest this please

**Pull request merged**